corehq/ex-submodules/pillowtop/utils.py

```diff
@@ -237,30 +249,33 @@ def _is_deleted(change):
             return bool(change.id)
 
     for change in changes:
-        action = {
-            "_index": index_info.alias,
-            "_id": change.id
-        }
-        if settings.ELASTICSEARCH_MAJOR_VERSION == 2:
-            action.update({"_type": index_info.type})
-
         if _is_deleted(change):
-            action.update({"_op_type": "delete"})
+            payload.append(BulkActionItem.delete_id(change.id))
         elif not change.deleted:
             try:
                 doc = change.get_document()
                 doc = doc_transform(doc)
-                action.update({
-                    "_op_type": "index",
-                    "_id": doc['_id'],
-                    "_source": doc
-                })
+                payload.append(BulkActionItem.index(doc))
             except Exception as e:
                 if not error_collector:
                     raise
                 error_collector.add_error(ChangeError(change, e))
-                continue
-        payload.append(action)
+        else:
+            # NOTE: Uncertain: this scenario seems possible because the 'if' and
+            # 'elif' clauses are checking different things (albeit, seemingly
+            # related).
+            #
+            # The previous implementation didn't have an 'else' clause here,
+            # meaning if this did happen, then an action lacking an '_op_type'
+            # key was added to the payload . I assume Elastic would return an
```

Comment:
Leaving the existing behavior will be tricky since the new BulkActionItem design is quite robust and doesn't support the caller "tweaking" the bulk payload in this way.  I'll do some testing and digging into these assumptions to determine exactly what happens in this scenario with the existing behavior. Maybe once I know that it will be more obvious how the existing behavior can be maintained.

Reply:
This is worse than I originally concluded.  If it is possible for an `else` condition to occur in the existing implementation, then it is resulting in a non-error Elasticsearch bulk action that is certainly not what we want.

Existing code [here](https://github.com/dimagi/commcare-hq/blob/a3ea4555770/corehq/ex-submodules/pillowtop/utils.py#L240-L263).

If the neither `if` or `elif` conditions are met, the result is an action being added to the payload that looks like this:

```python
{
    "_index": index_info.alias,
    "_type": index_info.type,
    "_id": change.id,
}
```

The [Elasticsearch bulk docs](https://elasticsearch-py.readthedocs.io/en/2.4.0/helpers.html#bulk-helpers) specify that an action lacking the `_op_type` key defaults to an `index` action.  This means that the above, sparse bulk action item will  always perform an `index` action for document `_id = change.id`. This will do one of:

1. index a new, empty (i.e. `_source = {}`) document which may or may not be using a valid `_id`, or
2. clear all properties on an existing document with that `_id`.

I tested both scenarios using the `bulk()` function and confirmed this is indeed what happens.

Changing the new BulkActionItem design to replicate this behavior is almost certainly not what we want to do. Thoughts?

Reply:
I submitted #31243 to fix this in order to keep this adapter PR free of Elastic processing logic changes.